CMakeLists.txt

```diff
@@ -321,7 +321,8 @@ if(NOT MSVC)
   set(CMAKE_REQUIRED_FLAGS "-msse4.2 -mpclmul")
 endif()
 
-CHECK_CXX_SOURCE_COMPILES("
+if (NOT PORTABLE OR FORCE_SSE42)
```

Comment:
@ajkr @pdillinger I believe this change makes PORTABLE builds on MSVC never use the CRC32c three way implementation even if the processor features are detected at runtime in crc32c.cpp's `Choose_Extend`, because `HAVE_SSE42` is now never defined and the runtime checks in `isPCLMULQDQ` and `isSSE42` are never invoked. If true, that's a regression from the previous behavior of PORTABLE builds.

Reply:
PORTABLE builds can test for POPCNT support using the CPUID instruction in a similar fashion to `isPCLMULQDQ` and `isSSE42` (POPCNT is bit 23 in the 3rd integer, similarly to how SSE 4.2 is bit 20).

Reply:
Runtime checking for processor features is not a sustainable model for us to implement. If you want peak performance, compile for your hardware. Also, at least on newer hardware, XXH3 checksum is faster than CRC32c.

I'm pretty sure that the runtime detection "feature" you allude to was an illusion anyway, because if -march= supports some feature, it is free to be used in code generation, without a runtime check, to optimize various patterns and APIs. If -march= does not support some feature, the compiler will refuse to compile a call to an intrinsic that requires that feature. Only __asm__ will bypass that check, which was not used in the x86 implementation of 3-way crc32c.

Reply:
The scenario I'm referring to is MSVC, where `-march` doesn't apply. You can build in MSVC without having compiler generated code use e.g., SSE 4 instructions, but still be able to invoke SSE 4 intrinsics - this is what MSVC PORTABLE builds did after a runtime decision in three way CRC32C as far as I can tell.
Our use case is shipping RocksDB as part of an end user Windows product that's installed on a variety of systems - unfortunately some of them with "antique" CPUs without POPCNT and some with the latest and greatest ones. We can't use non-PORTABLE builds due to the heterogeneous nature of end user devices. Building multiple RocksDB libraries with different build flags and selecting the one appropriate for a specific target system is not really viable. So, the previous state of affairs where PORTABLE builds dynamically selected optimized code paths at runtime when it had the most value (like block verification) was ideal.

Reply:
@pdillinger considering the MSVC will only use explicitly invoked intrinsics, do you think we can allow compilation when there are complete runtime checks on the features used? I am not planning to implement it but wondering if we would accept such a contribution. 

BTW, the previous code did not have complete runtime checks, in particular popcnt was not checked and still isn't, so I am not sure how the previous state was ideal.

Reply:
The previous code used POPCNT unconditionally at runtime in PORTABLE builds and that was definitely a bug. My only complaint is with the side effect of this PR that PORTABLE builds no longer attempt to use three way CRC32C on MSVC.
If you are inclined to accept such a contribution I'll go ahead and open a PR to address this.

Reply:
If MSVC works that way, sure send a PR.

Reply:
But we should bring clarity to what things like FORCE_SSE42 and HAVE_SSE42 really mean, or rename them or add new things that mean what we want. I had assumed that HAVE_* historically meant that the feature was available without a runtime check when I added HAVE_AVX2. Perhaps it historically (unclearly) meant that a runtime check was needed.